relentless/optimize/method.py

```diff
@@ -0,0 +1,83 @@
+import abc
+
+from .objective import ObjectiveFunction
+
+class Optimizer(abc.ABC):
+
+    @abc.abstractmethod
+    def optimize(self, objective):
+        pass
+
+class SteepestDescent(Optimizer):
+    def __init__(self, alpha, max_iter, abs_tol):
+        super().__init__()
+        self._alpha = alpha
+        self._max_iter = max_iter
+        self._abs_tol = abs_tol
+
+    def optimize(self, objective):
+        dvars = objective.design_variables()
+        if len(dvars) == 0:
+            return False
+
+        iter_num = 0
+        converged = False
+        while not converged and iter_num < self.max_iter:
+            res = objective.compute()
+            converged = self._has_converged(dvars, res)
+            for x in dvars:
+                x.value -= self.alpha*res.gradient(x)
+            iter_num += 1
+
+        return converged
+
+    def _has_converged(self, dvars, res):
+        for x in dvars:
+            grad = res.gradient(x)
+            try:
+                if abs(grad) > self.abs_tol:
+                    return False
+            except TypeError:
+                if abs(grad) > self.abs_tol[x]:
+                    return False
+        return True
+
+    @property
+    def alpha(self):
```

Comment:
I didn't notice this before, but you do want to call this something more descriptive like `step` or `step_size`? `alpha` is very unclear without seeing a mathematical formula.